File: core/src/main/java/com/nfthub/core/logger/LoggerFilter.java
package com.nfthub.core.logger;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.nfthub.core.exception.ExceptionResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.util.ContentCachingResponseWrapper;

import javax.servlet.*;
import javax.servlet.annotation.WebFilter;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@WebFilter(urlPatterns = "/*")
public class LoggerFilter implements Filter {
    public static final String USER_IP_HEADER = "X-FORWARDED-FOR";
    public static final String USER_AGENT_HEADER = "user-agent";
    public static final Logger logger = LoggerFactory.getLogger(LoggerFilter.class.getName());
    static final ObjectMapper objectMapper = new ObjectMapper();

    public static String getUserIp(HttpServletRequest request) {
        if (request.getHeader(USER_IP_HEADER) == null) {
            return request.getHeader(USER_IP_HEADER);
        }
        return request.getRemoteAddr();
    }

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain) throws IOException, ServletException {
        if (!(servletRequest instanceof HttpServletRequest) || !(servletResponse instanceof HttpServletRequest)) {
            chain.doFilter(servletRequest, servletResponse);
            logger.error("request or response is not HttpServletRequest");
            return;
        }

        HttpServletRequest request = (HttpServletRequest) servletRequest;
        HttpServletResponse response = (HttpServletResponse) servletResponse;
        ContentCachingResponseWrapper cachingResponseWrapper = new ContentCachingResponseWrapper(response);
        chain.doFilter(request, cachingResponseWrapper);
        logging(request, cachingResponseWrapper);
    }

    private void logging(HttpServletRequest request, ContentCachingResponseWrapper response) {
        if (response.getStatus() < 400) {
            infoLogging(request, response);
            return;
        }

        if (response.getStatus() != 500) {
            exceptionLogging(request, response);
        }

        // 500번은 ExceptionController 로깅
    }

    private void infoLogging(HttpServletRequest request, HttpServletResponse response) {
        InfoLog infoLog = new InfoLog();
        infoLog.setMethod(request.getMethod());
        infoLog.setResStatus(response.getStatus());
        infoLog.setRequestUrl(request.getRequestURI());
        infoLog.setUserIp(getUserIp(request));
        infoLog.setUserAgent(request.getHeader(USER_AGENT_HEADER));
        logger.info(infoLog.serialize());
    }

    private void exceptionLogging(HttpServletRequest request, ContentCachingResponseWrapper response) {
        ErrorLog errorLog = new ErrorLog();
        errorLog.setMethod(request.getMethod());
        errorLog.setResStatus(response.getStatus());
        errorLog.setRequestUrl(request.getRequestURI());
        errorLog.setUserIp(getUserIp(request));
        errorLog.setUserAgent(request.getHeader(USER_AGENT_HEADER));
        errorLog.setException(getExceptionResponse(response).getException());
        logger.error(errorLog.serialize());
    }

    private ExceptionResponse getExceptionResponse(ContentCachingResponseWrapper response) {
        ExceptionResponse exceptionResponse;
        try {
            String exceptionResponseString = new String(response.getContentAsByteArray(), response.getCharacterEncoding());
            exceptionResponse = objectMapper.readValue(exceptionResponseString, ExceptionResponse.class);
        } catch (Exception e) {
            exceptionResponse = new ExceptionResponse();
            exceptionResponse.setMessage(e.getLocalizedMessage());
            exceptionResponse.setException(e.getClass().getName());
        }
        return exceptionResponse;
    }
}

File: src/main/kotlin/com/nfthub/api/logger/LoggerFilter.kt
package com.nfthub.api.logger

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.nfthub.api.controller.ExceptionController
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.web.util.ContentCachingResponseWrapper
import javax.servlet.Filter
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.annotation.WebFilter
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@WebFilter(urlPatterns = ["/*"])
class LoggerFilter : Filter {
    companion object {
        val logger: Logger = LoggerFactory.getLogger("com.nfthub.api")
        const val userIdHeader = "X-FORWARDED-FOR"
        const val userAgentHeader = "user-agent"
        val mapper = jacksonObjectMapper()
    }

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        if (request !is HttpServletRequest || response !is HttpServletResponse) {
            chain.doFilter(request, response)
            logger.error("request or response is not HttpServletRequest")
            return
        }

        val resWrapper = ContentCachingResponseWrapper(response)
        chain.doFilter(request, resWrapper)

        logging(request, resWrapper)
        resWrapper.copyBodyToResponse()
    }

    private fun logging(req: HttpServletRequest, resWrapper: ContentCachingResponseWrapper) {
        if (resWrapper.status < 400) {
            logger.info(
                InfoLog(
                    resStatus = resWrapper.status,
                    method = req.method,
                    requestUrl = req.requestURL.toString(),
                    userIp = req.getHeader(userIdHeader) ?: req.remoteAddr,
                    userAgent = req.getHeader(userAgentHeader)
                ).serialize()
            )
            return
        }


        if (resWrapper.status < 500) {
            val errorResponse = resWrapper.getErrorResponse()
            logger.warn(
                ErrorLog(
                    resStatus = resWrapper.status,
                    method = req.method,
                    requestUrl = req.requestURI.toString(),
                    userIp = req.getHeader(userIdHeader) ?: req.remoteAddr,
                    userAgent = req.getHeader(userAgentHeader),
                    message = errorResponse.message,
                    exception = errorResponse.exception
                ).serialize()
            )
            return
        }

        if (resWrapper.status > 500) {
            val errorResponse = resWrapper.getErrorResponse()
            logger.error(
                ErrorLog(
                    resStatus = resWrapper.status,
                    method = req.method,
                    requestUrl = req.requestURI.toString(),
                    userIp = req.getHeader(userIdHeader) ?: req.remoteAddr,
                    userAgent = req.getHeader(userAgentHeader),
                    message = errorResponse.message,
                    exception = errorResponse.exception
                ).serialize()
            )
            return
        }

        if (resWrapper.status == 500)
            resWrapper.loggingInternalServerError(req)
    }

    private fun ContentCachingResponseWrapper.loggingInternalServerError(req: HttpServletRequest) =
        let {
            val errorResponseString = String(it.contentAsByteArray, charset(it.characterEncoding))
            try {
                // 파싱이 제대로 된다면 ExceptionController 처리
                mapper.readValue(errorResponseString, ExceptionController.ErrorResponse::class.java)
            } catch (e: Exception) {
                logger.error(
                    InternalServerErrorLog(
                        resStatus = it.status,
                        method = req.method,
                        requestUrl = req.requestURI.toString(),
                        userIp = req.getHeader(userIdHeader) ?: req.remoteAddr,
                        userAgent = req.getHeader(userAgentHeader),
                        message = e.message,
                        exception = e.javaClass.name,
                        errorStack = e.stackTraceToString().slice(0..500)
                    ).serialize()
                )
            }
        }

    private fun ContentCachingResponseWrapper.getErrorResponse(): ExceptionController.ErrorResponse = let {
        val errorResponseString = String(it.contentAsByteArray, charset(it.characterEncoding))
        return try {
            mapper.readValue(errorResponseString, ExceptionController.ErrorResponse::class.java)
        } catch (e: Exception) {
            ExceptionController.ErrorResponse(message = e.localizedMessage, exception = e.javaClass.toString())
        }
    }


}
